homeassistant/components/youless/sensor.py

```diff
@@ -0,0 +1,213 @@
+"""The sensor entity for the Youless integration."""
+from datetime import timedelta
+from typing import Any, Dict, Optional
+
+from youless_api import YoulessAPI
+from youless_api.youless_sensor import YoulessSensor
+
+from homeassistant import core
+from homeassistant.components.youless import DOMAIN
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_DEVICE, DEVICE_CLASS_POWER
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.typing import StateType
+from homeassistant.util import Throttle
+
+
+async def async_setup_entry(
+    hass: core.HomeAssistant, entry: ConfigEntry, async_add_entities
+):
+    """Initialize the integration."""
+    gateway = hass.data[DOMAIN][entry.entry_id]
+    device = entry.data[CONF_DEVICE]
+    sensors = [
+        GasSensor(gateway, device),
+        PowerMeterSensor(gateway, device, "low"),
+        PowerMeterSensor(gateway, device, "high"),
+        PowerMeterSensor(gateway, device, "total"),
+        CurrentPowerSensor(gateway, device),
+        DeliveryMeterSensor(gateway, device, "total"),
+        DeliveryMeterSensor(gateway, device, "current"),
+    ]
+
+    async_add_entities(sensors)
+
+
+class YoulessBaseSensor(Entity):
+    """The base sensor for Youless."""
+
+    def __init__(
+        self, gateway: YoulessAPI, device: str, friendly_name: str, sensor_id: str
+    ):
+        """Create the sensor."""
+        self._gateway = gateway
+        self._friendly_name = friendly_name
+        self._device = device
+        self._sensor_id = sensor_id
+
+    @Throttle(timedelta(seconds=30))
+    def update(self) -> None:
+        """Update the gateway value."""
+        self._gateway.update()
+
+    @property
+    def get_sensor(self) -> Optional[YoulessSensor]:
+        """Property to get the underlying sensor object."""
+        return None
+
+    @property
+    def unit_of_measurement(self) -> Optional[str]:
+        """Return the unit of measurement for the sensor."""
+        if self.get_sensor is None:
+            return None
+
+        return self.get_sensor.unit_of_measurement
+
+    @property
+    def state(self) -> StateType:
+        """Determine the state value, only if a sensor is initialized."""
+        if self.get_sensor is None:
+            return None
+
+        return self.get_sensor.value
+
+    @property
+    def unique_id(self) -> Optional[str]:
+        """Return the uniquely generated id."""
+        return f"{DOMAIN}_{self._device}_{self._sensor_id}"
```

Comment:
The device is a user-configured string, and thus not viable for use with a unique ID.

<https://developers.home-assistant.io/docs/entity_registry_index/#unique-id-requirements>

Reply:
I will have to look into if I can get some type of unique id from the Youless device, for the LS120 that should not be a problem as this exposes its MAC address using an API. For the LS110 this might be more of a challenge since this device does not.